crates/ruff_linter/src/rules/ruff/rules/none_not_at_end_of_union.rs

```diff
@@ -69,9 +78,37 @@ pub(crate) fn none_not_at_end_of_union<'a>(checker: &mut Checker, union: &'a Exp
         return;
     }
 
-    for none_expr in none_exprs {
-        checker
-            .diagnostics
-            .push(Diagnostic::new(NoneNotAtEndOfUnion, none_expr.range()));
+    for (pos, none_expr) in none_exprs.iter().with_position() {
+        let mut diagnostic = Diagnostic::new(NoneNotAtEndOfUnion, none_expr.range());
+        if matches!(pos, Position::Last | Position::Only) {
+            let mut elements = non_none_exprs
+                .iter()
+                .map(|expr| checker.locator().slice(expr.range()).to_string())
+                .chain(std::iter::once("None".to_string()));
+            let fix = if let Expr::Subscript(ast::ExprSubscript { slice, .. }) = union {
+                let applicability = if checker.comment_ranges().intersects(slice.range()) {
+                    Applicability::Unsafe
+                } else {
+                    Applicability::Safe
+                };
+                Fix::applicable_edit(
+                    Edit::range_replacement(elements.join(", "), slice.range()),
+                    applicability,
+                )
+            } else {
+                let applicability = if checker.comment_ranges().intersects(union.range()) {
+                    Applicability::Unsafe
+                } else {
+                    Applicability::Safe
+                };
+                Fix::applicable_edit(
+                    Edit::range_replacement(elements.join(" | "), union.range()),
+                    applicability,
+                )
+            };
+            diagnostic.set_fix(fix);
```

Comment:
Should we set the fix to all the diagnostics?

Reply:
Can you tell me a bit more why you think we should (or shouldn't) set the fix for all violations?

Reply:
@MichaReiser For example, this part looks like the same fix is set to all the diagnostics:

https://github.com/astral-sh/ruff/blob/8b9c843c727de463efccafc7e012414b93458dc7/crates/ruff_linter/src/rules/flake8_pyi/rules/duplicate_union_member.rs#L126-L129

Reply:
I'm actually leaning towards only emitting a single diagnostic rather than emitting multiple diagnostics. This should also simplify the code a bit because we can change `none_exprs` to an `Option` only tracking the `last_none` rather than all `None` values